src/Dialogs/Imap/ImapDialog.vala

```diff
@@ -253,23 +252,22 @@ public class OnlineAccounts.ImapDialog : Gtk.Window {
 
         login_page.cancel.connect (destroy);
 
-        login_page.next_button.clicked.connect (() => {
-            leaflet.visible_child = main_box;
+        main_page.shown.connect (() => {
             default_widget = save_button;
         });
 
-        save_page.close.connect (destroy);
-
-        save_page.back.connect (() => {
-            leaflet.navigate (Adw.NavigationDirection.BACK);
-            default_widget = save_button;
+        login_page.next_button.clicked.connect (() => {
+            push_page (main_page);
```

Comment:
AFAICT main_page will never be removed so its lifetime depends on `this` creating the ref cycle. To get around this you could just add a private Adw.NavigationPage main_page for it.